core/appmodule/genesis.go

```diff
@@ -0,0 +1,25 @@
+package appmodule
+
+import (
+	"context"
+
+	"cosmossdk.io/core/genesis"
+)
+
+// HasGenesis is the extension interface that modules should implement to handle
+// genesis data and state initialization.
+type HasGenesis interface {
+	AppModule
+
+	// DefaultGenesis writes the default genesis for this module to the target.
+	DefaultGenesis(genesis.Target) error
```

Comment:
I personally find these genesis.Target and .Source a bit confusing: who reads, who writes etc.

How about doing 2 interfaces? One simple HasGenesis with proto.Message that caters for 90% of cases, and one HasStreamingGenesis for large genesis files. The second one's method names would have "Streaming" inside.

Reply:
I don't have a strong preference between this approach and 2 interfaces. However, I do think we should consider streaming the 90% use case going forward and what modules that use collections or ORM will use. Proto genesis or raw json would be for legacy support IMHO

Reply:
I see. In this case would it make sense to only include one `HasGenesis` with streaming in core? Then, the raw json or proto genesis are runtime (v1) features, since they are legacy.

Reply:
That would be fine with me. I wonder what others think

Reply:
@alexanderbez @amaurym I've simplified this API to only support streaming, with the existing genesis being dealt with as legacy using the existing AppModule interfaces. Can you re-review?